src/ray/common/client_connection.cc

```diff
@@ -154,27 +154,22 @@ void ServerConnection<T>::DoAsyncWrites() {
     }
   }
   auto this_ptr = this->shared_from_this();
-  boost::asio::async_write(
-      ServerConnection<T>::socket_, message_buffers,
-      [this, this_ptr, num_messages](const boost::system::error_code &error,
-                                     size_t bytes_transferred) {
-        ray::Status status = ray::Status::OK();
-        if (error.value() != boost::system::errc::errc_t::success) {
-          status = boost_to_ray_status(error);
-        }
-        // Call the handlers for the written messages.
-        for (int i = 0; i < num_messages; i++) {
-          auto write_buffer = std::move(async_write_queue_.front());
-          write_buffer->handler(status);
-          async_write_queue_.pop_front();
-        }
-        // We finished writing, so mark that we're no longer doing an async write.
-        async_write_in_flight_ = false;
-        // If there is more to write, try to write the rest.
-        if (!async_write_queue_.empty()) {
-          DoAsyncWrites();
-        }
-      });
+  std::weak_ptr<ServerConnection<T>> pointer(this_ptr);
+  boost::asio::async_write(ServerConnection<T>::socket_, message_buffers,
```

Comment:
You're saying that when the write leads to a broken pipe error, that the callback (that is, the 3rd argument to `async_write` will not get called and it is leaked? Is this a bug in boost::asio?

Reply:
Looks like a bug in boost asio. Our observation is that when a remote raylet is killed, and this raylet is trying forward a task to that raylet. Sometimes the callback won't get called, even if we've explicitly call the callbacks in the destructor (#3572). It seems that boost is holding the lambda and never release it.

Reply:
It's worth checking if it's been fixed in a more recent boost version.

Reply:
I did some investigation on this, in this case we cannot rely on `write` to detect on broken connection, we need to use `read`. We might be able to use broken pipe error as well, but this is probably delayed because usually write would probably return before the data is actually sent to remote host, so most likely we cannot detect on time. `Read` should be able to detect immediately in this case. 

The underlying problem is related to [TCP half open connection](http://www.pcvr.nl/tcpip/tcp_conn.htm), which can happen when remote host crashes thus the TCP connection is not going through normal shutdown process. In this case the connection is "half opened", in that after the remote host crashes, the remote has closed its connection side indicating it's not sending, but we still might be able to send since the direction from us to remote is not closed... in this case if we read, we can detect the scenario immediately.

As for ASIO, this is probably a "feature" instead of a "bug", agreed this is a bit nasty though. Actually I found people have been discussing this [back in 2007.](http://boost.2283326.n4.nabble.com/ASIO-Detecting-broken-connections-td2569211.html)

Reply:
That's really interesting, thanks for looking into that!